File: src/main/java/com/brett/lec04/Lec04Main.java
package com.brett.lec04;

public class Lec04Main {
    public static void main(String[] args) {
//        compareTo();
//        equals();
        plus();
    }
    public static void plus(){
        JavaMoney javaMoney01 = new JavaMoney(2_000L);
        JavaMoney javaMoney02 = new JavaMoney(1_000L);
        System.out.println(javaMoney01.plus(javaMoney02));  // JavaMoney{amount=3000}
    }
    public static void equals(){
        JavaMoney javaMoney01 = new JavaMoney(1_000L);
        JavaMoney javaMoney02 = javaMoney01;
        JavaMoney javaMoney03 = new JavaMoney(1_000L);

        System.out.println(javaMoney01 == javaMoney02); // true
        System.out.println(javaMoney01 == javaMoney03); // false
        System.out.println(javaMoney01.equals(javaMoney03)); // 그냥 equals를 사용할 경우 false
        System.out.println(javaMoney01.equals(javaMoney03)); // JavaMoney에 equals를 구현하면 true
    }
    public static void compareTo(){
        JavaMoney javaMoney01 = new JavaMoney(2_000L);
        JavaMoney javaMoney02 = new JavaMoney(1_000L);

        if(javaMoney01.compareTo(javaMoney02) > 0){
            System.out.println("javaMoney01이 javaMoney02보다 금액이 큽니다.");
        }
    }
}

File: src/main/kotlin/com/brett/lec04/Lec04Main.kt
package com.brett.lec04

fun main(){
//    compareTo()
//    equals()
//    lazy()
    kotlinMoney()

}
fun kotlinMoney(){
    val money01 = Money(2_000L)
    val money02 = Money(1_000L)
    println(money01 + money02)  // Money(amount=3000)
}

fun lazy(){
    if(fun1() || fun2()){
        println("본문")
    }
    if(fun2() && fun1()){
        println("본문1")
    }
}
fun fun1() : Boolean{
    println("fun 1")
    return true;
}
fun fun2() : Boolean{
    println("fun 2")
    return false;
}
fun equals(){
    val money1 = JavaMoney(1_000L)
    val money2 = money1
    val money3 = JavaMoney(1_000L)

    println(money1 === money2) // true
    println(money1 === money3) // false
    /**
     * == 은
     * JavaMoney 의 equals로 처리됨
     */
    println(money1 == money3) // true
}
fun compareTo(){
    val money01 = JavaMoney(2_000L)
    val money02 = JavaMoney(1_000L)

    if(money01 > money02){
        println("javaMoney01이 javaMoney02보다 금액이 큽니다.")
    }
}
